### ISSUE TITLE
Change Baseline in TreeExplainer

### ISSUE THREAD

**Issue by BrunoBarre** [shap/shap#1253]
Hi all, I am working on a binary classification problem, using xgboost.

Once I trained my model, on X_train, I do the following line to get the SHAP model "trained" as well :
`tree_shap_explainer = shap.TreeExplainer( model, data=X_train, feature_perturbation="interventional", model_output="probability" )`

Then with tree_shap_explainer.expected_value, I get 0.38. Thus, for a observation that has the prediction 42%, it will explain the +4%. But this obs is in fact (with a 50% threshold) classified as  negative. 

Is there a way to explain the 0.42-0.5=-8%? Which would be more intuitive to understand, especially for people not working in the machine learning field?

I naively tried tree_shap_explainer.expected_value = 0.5, but of course it doesn't work this way... 
So my question is, could we change the baseline of the shap.TreeExplainer? Or, is there an even simpler trick to get a more intuitive explanation of the impact of every variable given by SHAP, relative to this 50% threshold? Thanks.

**Comment by InterferencePattern** [comment]
The way to think about the baseline is the expected value for the whole training dataset. (i.e. if the mean prediction of your model, based on that dataset, is 38%, then the expected or baseline value for your model would be .38. If the model only produces 0 or 1 and not probabilities, this means your predictions are 38% positive-class). Therefore, this is what your model considers to be the "typical" prediction, and all other predictions can be explained with respect to that number.

**Comment by BrunoBarre** [comment]
Yes I know that the baseline represents the mean prediction of my model on my whole train dataset. 

And yes my model output is the probabilty of belonging to the positive class.

But could I change that baseline to any probability by myself? For example to 50%, so it's eaysier for people to understand. 

Because if the baseline of 30% and a prediction of 40%, I think it's better to explain the 40% compared to 50%, since it gives insight about what variables make the obs not being predicted to the positive class, rather than explain the +10% compared to 30%, since it would give a positive sum of all impacts (10%), which is confusing because the prediction (in probability) is less than 50%, and most people will interpret it as not belonging to positive class.

I don't know if I am clear enough, or if that's even possible?

**Comment by BrunoBarre** [comment]
I upvote this question since it would be of great interest to me.

Would it be possible to manually select the baseline (ie the .expected_value) for SHAP approximation?

It would consist in including a parameter (for KernelShap, TreeShap, ...) like "baseline_value", that would be the expected value of the model used when it's None, and the baseline selected if it's a float, or a list of floats for multi-class case?
Thus, for a given instance/observation, the sum of the SHAP values would be equal to the difference between the model prediction for that instance and the selected baseline.

Concerning the real Shapley values, it is of course possible, to choose any baseline we want.

I don't know if I am clear enough, but don't hesitate to ask me question if needed :+1: 

Thanks!

**Comment by jaxondk** [comment]
I don't believe that is something you actually would want to do. It seems to me that you are confusing the baseline or expected value with the class threshold. 

The class threshold is typically 0.50, which is what you're suggesting setting the expected value to so that "it's easier for people to understand". But the expected value is not something you can just set and does not represent the division between classes. It is instead the expectation (or the mean) output of your model (which means there is an expected value for each class, as your model outputs something - i.e. the logits, typically - for every class), which is calculated off of your training data (typically a subset of your training data, or else calculating shap values takes unreasonably long). So your training data has an expected value that is very important and cannot just be arbitrarily set to something else. 

I hope that's helpful!

**Comment by BrunoBarre** [comment]
Hello and thanks for your answer.

That's exactly I want to do. To be able to change the baseline. In SHAP it is set as the expected value of the model. But in theoriginal Shapley value method (the "complete" one, not the SHAP approximation), you can set any baseline you want.

I did not say at all that setting the baseline as the expected value of the model is bad in any way, on the contrary it is mostly the kind of baseline I would choose. But it does make sense for some application to choose another baseline value.

As far as I understand, SHAP made the choice to set the expected value as the baseline, and it is indeed the best choice most of the time.

But I want to know if by any means, there is a way to choose my own baseline( like 50%). Right now, it is not possible, but is this a just a simple "problem" of implementation (it seems that the baseline is set, harcoded, as the expected value here), or is that incompatible with the SHAP method -which is an approximation of actual Shapley values?

Thanks.

**Comment by XingLiu0923** [comment]
I also came across a similar problem recently, and would like to know if there is any way to adjust the baseline manully. Sigh..

**Comment by vciulei-w** [comment]
Hello, I'm also interested in this issue and I believe it can be quite useful. @BrunoBarre you're right, in the original Shapley paper, the baseline can be changed, and it can be a powerful tool to answer the question: 'how much is this feature contributing to the prediction relative to a specific subset of the training dataset?' @jimbudarz maybe this is more clear? For example I want to know how much the price of a house is predicted to increase compared only to the average predictions of houses with 3 rooms (or some other subset). It could also work for binary classifications.

I'm still new to SHAP and don't know the in's & out's yet, but perhaps one can choose the specific subset they're interested in and run SHAP in order to get the expected prediction of that subset, and then somehow use the same explainer object (thinking out loud here) to keep the initial baseline and compare it with SHAP values from the other/entire subset your're interested in.

@slundberg I know you probably get 100s of pings per day :) If by any chance you have to check this out, highly appreciate it.

**Comment by elisabaldassari** [comment]
Hi, has someone managed to change it in the end? I have tried to modify the source code (focused on the _tree.py and _explainer.py) to set the baseline (expected_value) to a different value, but it gets overwritten. 
I think that solving this issue can be quite useful :)

**Comment by drtobybrown** [comment]
+1 for this issue a year on to see if any progress has been made..

@BrunoBarre

**Comment by emilyybo** [comment]
I am looking to change the baseline in TreeExplainer as well. I would appreciate an update.

**Comment by drtobybrown** [comment]
The solution I found was to pass a background data set to `TreeExplainer` with the desired baseline

**Comment by emilyybo** [comment]
Thank you @drtobybrown

**Comment by emilyybo** [comment]
@drtobybrown I fed the TreeExplainer function with the same model and the same baseline data. However, I get different expected values every time I ran the following code. Do you have insight on why I don't get the same results? 

```

referencedat = {"var1":[1], "var2":[1],'var3':[2],  'var4':[1], 'var5':[1], 'var6':[1],'var7':[3] }

referencedf = pd.DataFrame(referencedat)

explainer = shap.TreeExplainer(m_xg, data= referencedf)

explainer.expected_value
```

**Comment by roy651** [comment]
+1 on this.
Trying to present "supporting" and "contradicting" attributes on a binary classification decision, I realize that the supporting/contradicting factors do not relate to the "direction" of the decision as a human is expected to understand it but rather it relates to the support/contradict per the mean baseline value... 
This is completely un-intuitive for human understanding and defies the entire purpose of providing an explanation. Please consider offering a manual override for this. In the meantime, guess I'll be using the above mentioned work-around :-/

**Comment by ananya-8451** [comment]
+1  for this issue. This can be very useful to us as well. 
'how much is this feature contributing to the prediction relative to a specific subset of the training dataset?'

I tried changing the background/reference dataset. That does solve the issue. However, that only works for feature_perturbation = 'interventional'. If we try to run it with feature_perturbation = tree_path_dependent (which is preferrable to interventional) it gives us this error:

AssertionError: The background dataset you provided does not cover all the leaves in the model, so TreeExplainer cannot run with the feature_perturbation="tree_path_dependent" option! Try providing a larger background dataset, no background dataset, or using feature_perturbation="interventional".

**Comment by AnuragIITM22** [comment]
Did anyone get the solution for the above issue? How can I change my baseline in SHAP based on my choice?

**Comment by kadambini-indurkar** [comment]
Do we have a solution here? How to change the baseline of SHAP?

**Comment by kadambini-indurkar** [comment]
When using tweedie objective function the contribution to other drivers are very less, the model I am using in this case is LGBM

**Comment by gialmisi** [comment]
Also interested in this.
